packages/molecule/src/utils.ts

```diff
@@ -0,0 +1,231 @@
+  if (data === null || data === undefined) throw new Error("NonNullable");
+}
+
+export function toMolTypeMap(results: MolType[]): MolTypeMap {
+  const map: MolTypeMap = {};
+  results.forEach((result) => {
+    map[result.name] = result;
+  });
+  return map;
+}
+
+export function validateParsedResults(
+  results: MolType[],
+  option: ParseOptions,
+) {
+  checkDuplicateNames(results);
+  // skip check is refs presents
+  if (!option.skipDependenciesCheck && !option.refs) {
+    checkDependencies(results);
+  }
+}
+
+function checkDuplicateNames(results: MolType[]) {
+  const names = new Set<string>();
+  results.forEach((result) => {
+    const currentName = result.name;
+    if (names.has(currentName)) {
+      throw new Error(`Duplicate name: ${currentName}`);
+    }
+    names.add(currentName);
+    const currentType = result.type;
+    // check duplicate field names in `struct` and `table`
+    if (currentType === "struct" || currentType === "table") {
+      const fieldNames = new Set<string>();
+      (result as Struct).fields.forEach((field: Field) => {
+        const currentFieldName = field.name;
+        if (fieldNames.has(currentFieldName)) {
+          throw new Error(`Duplicate field name: ${currentFieldName}`);
+        }
+        fieldNames.add(currentFieldName);
+      });
+    }
+  });
+}
+
+export function checkDependencies(results: MolType[]): void {
+  const map = toMolTypeMap(results);
+  for (const key in map) {
+    const molItem = map[key];
+    nonNull(molItem);
+    const type = molItem.type;
+    switch (type) {
+      case "array":
+      case "struct": {
+        assertFixedMolType(molItem.name, map);
+        break;
+      }
+      case "vector":
+      case "option": {
+        if ((molItem as Vector).item !== byte) {
+          nonNull(map[(molItem as Vector).item]);
+        }
+        break;
+      }
+      case "union": {
+        const unionDeps = molItem.items;
+        unionDeps.forEach((dep) => {
+          if (typeof dep === "string" && dep !== byte) {
+            nonNull(map[dep]);
+          }
+          if (Array.isArray(dep)) {
+            const [key, id] = dep;
+            // check if the id is a valid uint32
+            mol.Uint32.encode(id);
+            nonNull(map[key]);
+          }
+        });
+        break;
+      }
+      case "table": {
+        const tableDeps = molItem.fields.map((field: Field) => field.type);
+        tableDeps.forEach((dep: string) => {
+          if (dep !== byte) {
+            nonNull(map[dep]);
+          }
+        });
+
+        break;
+      }
+      default:
+        break;
+    }
+  }
+}
+
+/**
+ * mol type `array` and `struct` should have fixed byte length
+ */
+function assertFixedMolType(name: string, map: MolTypeMap): void {
+  const molItem = map[name];
+  nonNull(molItem);
+  const type = molItem.type;
+  switch (type) {
+    case "array": {
+      if (molItem.item !== byte) {
+        assertFixedMolType(molItem.name, map);
```

Comment:
I haven't actually run this code, but it looks like the implementation here will result in an infinite loop. Perhaps we should test for the case where the array item is not `byte`?